### PR TITLE
Problem: memiavl leave tmp directories behind

### PR DESCRIPTION
**Pull request by mmsqe**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

Closes: https://github.com/crypto-org-chain/cronos/issues/1095
# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db.go [hunk 1]
```diff
@@ -203,6 +203,13 @@ func Load(dir string, opts Options) (*DB, error) {
 		}
 	}
 
+	snapshotDir := snapshotName(db.lastCommitInfo.Version)
+	tmpDir := snapshotDir + "-tmp"
+	path = filepath.Join(db.dir, tmpDir)
+	if err := os.Remove(path); err != nil {
```
Comment (yihuang) [inline memiavl/db.go:209]: I think we need to `RemoveAll` here

Comment (mmsqe) [reply memiavl/db.go]: yea, but not sure if del with -tmp suffix will be heavy

Comment (yihuang) [reply memiavl/db.go]: should be alright, this is just the startup.

File: memiavl/db.go [hunk 2]
```diff
@@ -203,9 +203,56 @@ func Load(dir string, opts Options) (*DB, error) {
...
 }
 
+func removeAllTmpDirs(rootDir string) error {
+	const maxRemovals = 1000
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var errors []error
+	sem := make(chan struct{}, maxRemovals)
+
+	err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
```
Comment (yihuang) [inline memiavl/db.go:220]: `os.ReadDir()` is better, `Walk` will traverse recursively into the directories.

File: memiavl/db.go [hunk 3]
```diff
@@ -203,9 +203,57 @@ func Load(dir string, opts Options) (*DB, error) {
...
+			continue
+		}
+
+		if i >= maxRemovals {
+			break
+		}
+		i++
+
+		wg.Add(1)
+		go func(dirEntry os.DirEntry) {
```
Comment (yihuang) [inline memiavl/db.go:236]: Is `RemoveAll` so slow that need concurrency to optimize it?

Comment (mmsqe) [reply memiavl/db.go:236]: Just a limit since I'm not sure how many total removal to handle with

Comment (yihuang) [reply memiavl/db.go:236]: if it's trivial then don't need the extra code complication, and concurrency has it's own overhead.

Comment (yihuang) [reply memiavl/db.go:236]: Can I simplify this code?

Comment (mmsqe) [reply memiavl/db.go]: done, just think concurrent is better in for loop

**Review by yihuang — APPROVED** [review]

File: memiavl/db.go [hunk 4]
```diff
@@ -206,6 +212,31 @@ func Load(dir string, opts Options) (*DB, error) {
...
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		version, err := parseTmpVersion(entry.Name())
+		if err != nil {
+			continue
+		}
+		if version >= v {
```
Comment (yihuang) [inline memiavl/db.go:229]: Why do pruning here, and it ignores the keep-recent settings

**Review by yihuang — CHANGES_REQUESTED** [review]
Don't do pruning, just cleanup tmp dirs

File: memiavl/db.go (hunk 4 above)
Comment (yihuang) [reply memiavl/db.go:229]: I see the problem, we need to introduce readonly mode, only remove files in write mode.

**Review by yihuang — APPROVED** [review]

**Pull request merged**